docs/dev/codebase_walkthrough.rst

```diff
@@ -0,0 +1,227 @@
+=======================================
+**TVM Codebase Walkthrough by Example**
+=======================================
+
+Getting to know a new codebase can be a challenge. This is especially true for a codebase like that of TVM, where different components interact in non-obvious ways. In this guide, we try to illustrate the key elements that comprise a compilation pipeline with a simple example. For each important step, we show where in the codebase it is implemented. The purpose is to let new developers and interested users dive into the codebase more quickly.
+
+*******************************************
+Codebase Structure Overview
+*******************************************
+
+At the root of the TVM repository, we have following subdirectories that together comprise a bulk of the codebase.
+
+- ``src`` - C++ code for operator compilation and deployment runtimes.
+- ``python`` - Python frontend that wraps C++ functions and objects implemented in ``src``.
+- ``topi`` - Compute definitions and backend schedules for standard neural network operators.
+- ``nnvm`` - C++ code and Python frontend for graph optimization and compilation. Depends on three directories above.
+
+Using standard Deep Learning terminologies, ``nnvm`` is the component that manages a computational graph, and nodes in a graph are compiled and executed using infrastructures implemented in ``src`` and ``python``. Operators corresponding to each node are registered in ``nnvm``. Registration can be done via C++ or Python. Implementations for operators are in ``topi``, and they are also coded in either C++ or Python.
+
+When a user invokes graph compilation by ``nnvm.compiler.build(...)``, the following sequence of actions happens for each node in the graph:
+
+- Look up an operator implementation by querying the operator registry
+- Generate a compute expression and a schdule for the operator
+- Compile the operator into object code
+
+One of the interesting aspects of TVM codebase is that interop between C++ and Python is not unidirectional. Typically, all code that do heavy liftings are implemented in C++, and Python bindings are provided for user interface. This is also true in TVM, but in TVM codebase, C++ code also call into functions defined in a Python module. For example, the convolution operator is implemented in Python, and its implementation is invoked from C++ code in nnvm.
+
```

Comment:
Given that we are transitioning to relay? Is it possible to directly introduce the graph related part in the context of relay?

Reply:
Sure, I can update nnvm part with relay to reflect the latest development. But I think it is a good idea to leave some note about nnvm because many doc and tutorials are written in terms of nnvm. People may be confused If we don't mention nnvm in this doc.